hack/e2e.sh

```diff
@@ -56,40 +62,232 @@ while getopts "h?" opt; do
     h|\?)
         usage
         exit 0
-        ;;  
+        ;;
     esac
 done
 
 if [ "${1:-}" == "--" ]; then
     shift
 fi
 
+hack::ensure_kind
+hack::ensure_kubectl
+hack::ensure_helm
+
 DOCKER_REGISTRY=${DOCKER_REGISTRY:-localhost:5000}
 IMAGE_TAG=${IMAGE_TAG:-latest}
+CLUSTER=${CLUSTER:-tidb-operator}
+KUBECONFIG=${KUBECONFIG:-~/.kube/config}
+KUBECONTEXT=kind-$CLUSTER
 SKIP_BUILD=${SKIP_BUILD:-}
 SKIP_IMAGE_BUILD=${SKIP_IMAGE_BUILD:-}
+SKIP_UP=${SKIP_UP:-}
+SKIP_DOWN=${SKIP_DOWN:-}
+KIND_DATA_HOSTPATH=${KIND_DATA_HOSTPATH:-none}
+KUBE_VERSION=${KUBE_VERSION:-v1.12.10}
 
 echo "DOCKER_REGISTRY: $DOCKER_REGISTRY"
 echo "IMAGE_TAG: $IMAGE_TAG"
+echo "CLUSTER: $CLUSTER"
+echo "KUBECONFIG: $KUBECONFIG"
+echo "KUBECONTEXT: $KUBECONTEXT"
 echo "SKIP_BUILD: $SKIP_BUILD"
 echo "SKIP_IMAGE_BUILD: $SKIP_IMAGE_BUILD"
+echo "SKIP_UP: $SKIP_UP"
+echo "SKIP_DOWN: $SKIP_DOWN"
+echo "KIND_DATA_HOSTPATH: $KIND_DATA_HOSTPATH"
+echo "KUBE_VERSION: $KUBE_VERSION"
 
-if [ -n "$SKIP_BUILD" ]; then
-    echo "info: skip building images"
-    export NO_BUILD=y
-fi
+# https://github.com/kubernetes-sigs/kind/releases/tag/v0.6.0
+declare -A kind_node_images
+kind_node_images["v1.11.10"]="kindest/node:v1.11.10@sha256:44e1023d3a42281c69c255958e09264b5ac787c20a7b95caf2d23f8d8f3746f2"
+kind_node_images["v1.12.10"]="kindest/node:v1.12.10@sha256:e93e70143f22856bd652f03da880bfc70902b736750f0a68e5e66d70de236e40"
+kind_node_images["v1.13.12"]="kindest/node:v1.13.12@sha256:ad1dd06aca2b85601f882ba1df4fdc03d5a57b304652d0e81476580310ba6289"
+kind_node_images["v1.14.9"]="kindest/node:v1.14.9@sha256:00fb7d424076ed07c157eedaa3dd36bc478384c6d7635c5755746f151359320f"
+kind_node_images["v1.15.6"]="kindest/node:v1.15.6@sha256:1c8ceac6e6b48ea74cecae732e6ef108bc7864d8eca8d211d6efb58d6566c40a"
+kind_node_images["v1.16.3"]="kindest/node:v1.16.3@sha256:bced4bc71380b59873ea3917afe9fb35b00e174d22f50c7cab9188eac2b0fb88"
```

Comment:
Shall we iterate all the patch versions for each minor version? e.g. the caller may specify `v1.16.2`

Reply:
this is a list of pre-built kind node images. If `v1.16.2` is specified, the script will fail with an "unsupported version" error.